Convert code governance page to rst (#274)
mautic/mautic-community-handbook: docs/governance/code-governance.rst

Code governance
################

> Tiers for triaging pull requests

Reviewer: We need to add the word triaging to the accept.txt file for Vale, as it doesn't know of the word.

Author: > We need to add the word triaging to the accept.txt file for Vale, as it doesn't know of the word.

How do I do that? I cant find the file you mentioned(accept.txt )

Reviewer 2: <img width="319" alt="Screenshot 2025-03-15 at 19 28 58" src="https://github.com/user-attachments/assets/43c68437-b56d-4f42-b82c-0ccfda8557a5" />
You'll find the accept.txt on the `.github` folder

Author: Thank yoy

********************************

We've adopted a tier-based approach to categorising pull requests (PRs) in the Mautic Community. This allows the Product Team to establish clear governance approaches to how PRs will be dealt with and what the criteria will be for merging.

This allows the processing of minor fixes and improvements more rapidly, while time-boxing community testing for major pull requests, enabling the Product Team to make decisions on whether to merge new features once those time periods have passed.

Tier 1
======
Small changes such as typo fixing, bug fixes, translation changes, minor UI improvements or minor enhancements. Most simple bug fixes are probably going to be in Tier 1.

Examples (Tier 1)
------------------

> https://github.com/mautic/mautic/pull/8393 (simple bug fix with very few changes)

Reviewer: Links need to be created using the command `make link` at the command line in the docs folder, and then the link macro inserted here. Please can you do that for each link? Thanks! (more detail in the issue here: https://github.com/mautic/mautic-community-handbook/issues/219)

Author: Thank you for the feedback. I have noted that

Author: > Links need to be created using the command `make link` at the command line in the docs folder, and then the link macro inserted here. Please can you do that for each link? Thanks! (more detail in the issue here: #219)

Thank you for your feedback. But these links do not have a name particularly those referring to GitHub. How do I name them?

Author: > Links need to be created using the command `make link` at the command line in the docs folder, and then the link macro inserted here. Please can you do that for each link? Thanks! (more detail in the issue here: #219)

![pr](https://github.com/user-attachments/assets/9e2716ca-b4cf-4ded-a21f-071ba4e00c17)
Is the name ok for the first PR?

Reviewer: The name for the link really doesn't matter, it's just what's used within the link macro to render. The part that matters is what the visitor sees - for that I'd use the full URL in this case, because that's what we want to be rendered on the page as the anchor text.

For the filename I would check the existing structure and use something aligned - I guess something like github_pr_3893 - so that it's easy to understand what the links do from viewing the filename.

Author: Note thanks I will effect it

https://github.com/mautic/mautic/pull/8116/files (very simple bug fix)

Requirements for merging a Tier 1 PR
-------------------------------------
 

>  * Thorough testing to ensure that it does what it’s expected to do and does'nt break anything else

Reviewer: ```suggestion
 * Thorough testing to ensure that it does what it’s expected to do and doesn't break anything else
```
Typo fix here

Examples (Tier 3)
------------------

https://github.com/mautic/mautic/pull/6875 (complex code changes across multiple areas, with significant impact)

https://github.com/mautic/mautic/pull/6584 (not too complex but a large visual change in the UI with the potential to cause significant disruption to the end user)

Requirements for Merging a Tier 3 PR    
-------------------------------------
* Thorough testing to ensure that it does what it’s expected to do and does not break anything else
* Thorough testing across the whole product
* Full unit/functional test coverage
* Full documentation support
* Code review from 1-2 core team members
* Full testing and approval by at least two community members (could be the people doing the review, but should not be from the same company/organisation as the person submitting the PR as mentioned above)
* 2 week community review period - after which Product Team decides via a vote in the next available team meeting whether to merge (assuming code review and all other requirements are in place to permit merging)

.. note::

>     Tier 3 PRs which relate to new features that impact users of Mautic **should be linked to in an individual forum thread** at `forum.mautic.org/c/ideas <https://forum.mautic.org/>`_ and discussed on the forums. This gives the non-technical user base an opportunity to be involved in the discussions on new features. 

Reviewer: Please use the forum link macro here, that's how we do links. You'll find them in the links directory.

Author: > Please use the forum link macro here, that's how we do links. You'll find them in the links directory.

Noted. The forum link macro directs to the support page should I create another file for it to land the user to the main page? 

![forum](https://github.com/user-attachments/assets/a7f6844f-7cde-4640-b14d-6875a850d8a9)

Reviewer: Yes, if you need a link which goes somewhere else other than the support category, then I would recommend making a new link and using the filename to explain the distinction so that other writers will understand what it is linking to. However, think about whether it's necessary because a lot of the time folks need to go to the support section.

Author: Alright I would consider it

If a thread doesn’t exist prior to the PR being submitted (for example, a feature created without the community suggesting it), it should be created during triage when a Tier 3 & Feature label is added.

> The Github URL should be shared in a single line on the forum thread (enables the ‘onebox’ to be created) which will create a reciprocal link on the GitHub PR.

Reviewer: ```suggestion
The GitHub URL should be shared in a single line on the forum thread (enables the ‘onebox’ to be created) which will create a reciprocal link on the GitHub PR.
```

Tiers for triaging issues
*************************

The same tiered approach used in triaging pull requests is also used with issues. The triage team are responsible for daily/weekly/monthly and quarterly review processes.

Tier 1 issues
==============

Minor issues which are non-critical in nature.

Examples (Tier 1 issues)
-------------------------
https://github.com/mautic/mautic/issues/8974 - an issue with translations

https://github.com/mautic/mautic/issues/8986 - checkbox in the wrong place

Tier 2 issues
==============
Issues which are more complex and/or which may impact a large volume of users.

Examples (Tier 2 issues)
-------------------------

> https://github.com/mautic/mautic/issues/8621 - impacts all Sparkpost users

Reviewer: ```suggestion
https://github.com/mautic/mautic/issues/8621 - impacts all SparkPost users
```

:xref:`github_pr_7432`- a lot of new code which extends an existing feature, enhancing reports with a scheduling option

Requirements for merging a Tier 2 PR
--------------------------------------

* Thorough testing to ensure that it does what it’s expected to do and does not break anything else
* Full unit testing coverage
* Full documentation support if it’s a feature change/addition
* Code review from 1 core team member
* Full testing and approval by at least one community member (could be the person doing the code review, but should not be from the same company/organisation as the person submitting the PR as mentioned above)

Tier 3
======
Major changes touching multiple parts of Mautic, completely changing the way some aspect of the code works, or large amounts of code changed.

Examples (Tier 3)
------------------

:xref:`github_pr_6875`- complex code changes across multiple areas, with significant impact

:xref:`github_pr_6584`- not too complex but a large visual change in the UI with the potential to cause significant disruption to the end user

Requirements for Merging a Tier 3 PR    
-------------------------------------

> * Thorough testing to ensure that it does what it’s expected to do and does not break anything else

Reviewer: ```suggestion
* Thorough testing to ensure that it does what it’s expected to do and doesn't break anything else
```

* Thorough testing across the whole product
* Full unit/functional test coverage
* Full documentation support
* Code review from 1-2 core team members

> * Full testing and approval by at least two community members (could be the people doing the review, but should not be from the same company/organisation as the person submitting the PR as mentioned above)

Reviewer: ```suggestion
* Full testing and approval by at least two community members - could be the people doing the review, but shouldn't be from the same company/organisation as the person submitting the PR as mentioned above
```

* 2 week community review period - after which Product Team decides via a vote in the next available team meeting whether to merge (assuming code review and all other requirements are in place to permit merging)

.. note::
    Tier 3 PRs which relate to new features that impact users of Mautic **should be linked to in an individual forum thread** at :xref:`Ideas and Feature Requests` and discussed on the forums. This gives the non-technical user base an opportunity to be involved in the discussions on new features. 

> If a thread doesn’t exist prior to the PR being submitted (for example, a feature created without the community suggesting it), it should be created during triage when a Tier 3 & Feature label is added.

Reviewer: ```suggestion
If a thread doesn’t exist prior to the PR being submitted - for example, a feature created without the community suggesting it - it should be created during triage when a Tier 3 & Feature label is added.
```

This will also need converting to active voice. Vale hasn't picked it up yet.

> The GitHub URL should be shared in a single line on the forum thread (enables the ‘onebox’ to be created) which will create a reciprocal link on the GitHub PR.

Reviewer: ```suggestion
The GitHub URL should be shared in a single line on the forum thread - this enables the ‘onebox’ to be created - which will create a reciprocal link on the GitHub PR.
```

Please add onebox to the accept list.

Tiers for triaging issues
*************************

The same tiered approach used in triaging pull requests is also used with issues. The triage team are responsible for daily/weekly/monthly and quarterly review processes.

Tier 1 issues
==============

Minor issues which are non-critical in nature.

> Examples (Tier 1 issues)

Reviewer: ```suggestion
Examples - Tier 1 issues
```

-------------------------
:xref:`github_pr_8974` - an issue with translations

:xref:`github_pr_8986` - checkbox in the wrong place

Tier 2 issues
==============
Issues which are more complex and/or which may impact a large volume of users.

> Examples (Tier 2 issues)

Reviewer: ```suggestion
Examples - Tier 2 issues
```

-------------------------
:xref:`github_pr_8621`- impacts all SparkPost users

:xref:`github_pr_7062`- appears often but not always when multiple forms are added on a single page, making it tricky to reproduce

Tier 3 issues
==============
Complex issues which impact multiple areas of Mautic or which require extensive debugging to identify/resolve.

> Examples (Tier 3 issues)

Reviewer: ```suggestion
Examples - Tier 3 issues
```

-------------------------
:xref:`github_pr_9072`- will likely touch many areas and is complex to solve, requiring in-depth technical knowledge of PHPUnit

:xref:`github_pr_7032`- took a lot of time to confirm the issue and find a proper solution without breaking other things

Labels
=======
We currently have :xref:`mautic label` which are organised as below

> 

Reviewer: Think we can remove the extra spaces here?

Complexity-based labels
------------------------
* Tier 1-2-3 [T1, T2, T3] (for PR’s and issues)

Semantic versioning related labels
-----------------------------------
* BC Break
* Deprecation
* Essential (required to close the milestone)

Type-based labels
------------------
* Bug
* New feature
* Enhancement

> * Dependencies (used by Dependabot)

Reviewer: Please add Dependabot to the accept list. Thanks!

* Regression

Status-based labels

> --------------------

Reviewer: Please fix the underline here.

* Needs triage

> * Rebase needed

Reviewer: Please add Rebase to the accept list (with capital)

* WIP
* Requires automated tests
* Requires documentation
* Requires code review
* Pending feedback
* Pending code changes
* Has conflicts

> * Ready to test (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is mergeable)

Reviewer: ```suggestion
* Ready to test (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is considered ready for merging)
```

> * Pending test confirmation (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is mergeable)

Reviewer: ```suggestion
* Pending test confirmation (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is considered ready for merging)
```

> * Ready to commit (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written, has the required signoff/approvals and is mergeable)

Reviewer: ```suggestion
* Ready to commit (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written, has the required signoff/approvals and is considered ready for merging)
```

Area Affected Labels (which part of the product does this affect?)
===================================================================
* Assets
* Builders (email and LP)
* Calendar
* Campaigns
* Categories
* Channels
* Companies
* Configuration
* Contacts
* Dashboard
* Dynamic Content
* Editor
* Email

> * File Uploader

Reviewer: Please add Uploader to the accept list

Code governance
###############

.. toctree::
   :maxdepth: 2
   :hidden:

   deprecation_policy

> 

Reviewer 3: Error message:

`/workspace/mautic-community-handbook/docs/governance/code_governance.rst:4: WARNING: toctree contains reference to nonexisting document 'governance/deprecation_policy' [toc.not_readable]`

Above error is coming from these lines because we don't have any `deprecation_policy.rst` file yet. Please remove these.

Author: I did update it locally, forgot to push. Thank you so much

Tier 3
======
Major changes touching multiple parts of Mautic, completely changing the way some aspect of the code works, or large amounts of code changed.

Examples (Tier 3)
------------------

:xref:`github_pr_6875`- complex code changes across multiple areas, with significant impact

:xref:`github_pr_6584`- not too complex but a large visual change in the UI with the potential to cause significant disruption to the end user

Requirements for Merging a Tier 3 PR    
-------------------------------------
* Thorough testing to ensure that it does what it’s expected to do and does not break anything else
* Thorough testing across the whole product
* Full unit/functional test coverage
* Full documentation support
* Code review from 1-2 core team members
* Full testing and approval by at least two community members (could be the people doing the review, but should not be from the same company/organisation as the person submitting the PR as mentioned above)
* 2 week community review period - after which Product Team decides via a vote in the next available team meeting whether to merge (assuming code review and all other requirements are in place to permit merging)

.. note::

>     Tier 3 PRs which relate to new features that impact users of Mautic **should be linked to in an individual forum thread** at :xref:`Ideas and Feature Requests` and discussed on the forums. This gives the non-technical user base an opportunity to be involved in the discussions on new features. 

Reviewer: ```suggestion
    Tier 3 PRs which relate to new features that impact users of Mautic **should be linked to in an individual forum thread** at :xref:`Ideas and Feature Requests` and discussed on the forums. This gives the non-developer user base an opportunity to be involved in the discussions on new features. 
```

I prefer if we refer to these folks as 'non developer' rather than 'non technical' nowdays - many marketers are very technical ;)

.. note::
    Tier 3 PRs which relate to new features that impact users of Mautic **should be linked to in an individual forum thread** at :xref:`Ideas and Feature Requests` and discussed on the forums. This gives the non-developer user base an opportunity to be involved in the discussions on new features. 

If a thread doesn’t exist prior to the PR being submitted - for example, a feature created without the community suggesting it - it should be created during triage when a Tier 3 & Feature label is added.

The GitHub URL should be shared in a single line on the forum thread - this enables the ‘onebox’ to be created - which will create a reciprocal link on the GitHub PR.

Tiers for triaging issues
*************************

The same tiered approach used in triaging pull requests is also used with issues. The triage team are responsible for daily/weekly/monthly and quarterly review processes.

Tier 1 issues
==============

Minor issues which are non-critical in nature.

Examples - Tier 1 issues
-------------------------
:xref:`github_pr_8974` - an issue with translations

:xref:`github_pr_8986` - checkbox in the wrong place

Tier 2 issues
==============

> Issues which are more complex and/or which may impact a large volume of users.

Reviewer: No, we're talking the general mass of folks who use Mautic rather than the individual user account.

* Pending feedback
* Pending code changes
* Has conflicts
* Ready to test (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is mergeable)
* Pending test confirmation (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written and is mergeable)
* Ready to commit (PR’s only, and only applied when the PR is passing tests, has no conflicts, has automated tests written, has the required signoff/approvals and is mergeable)

Area Affected Labels (which part of the product does this affect?)
===================================================================
* Assets
* Builders (email and LP)
* Calendar
* Campaigns
* Categories
* Channels
* Companies
* Configuration
* Contacts
* Dashboard
* Dynamic Content
* Editor
* Email
* File Uploader
* Focus Items
* Forms
* Import Export
* Installation
* Integrations
* Landing Pages
* Notifications
* Plugin
* Points/Scoring
* Queue
* Reports
* Roles
* Segments
* SMS
* Social Monitoring
* Stages
* Tags
* Tracking
* Translations
* User Interface
* User Experience
* Webhooks
* Widgets

Some points of clarification
=============================

> **Core Team:** Individuals selected by the Project Lead with technical ability to manage and maintain the core of Mautic - includes Release Leaders, Core Committers, Maintainers (see :xref:`mautic.org/about/governance`). Currently listed here.

Reviewer: Looks like we're missing a link on the 'Currently listed here' bit, and the link should be https://mautic.org/governance/ (no /about/ any more)

Reviewer: This needs following up - link is still missing to the core team where it says 'here'.

Author: > This needs following up - link is still missing to the core team where it says 'here'.

Yeah the link in the old docs says it is broken. You said that will be fixed later when I made enquiry about it on Slack

Some points of clarification
============================
**Core Team:** Individuals selected by the Project Lead with technical ability to manage and maintain the core of Mautic - includes Release Leaders, Core Committers, Maintainers (see :xref:`mautic.org/about/governance`). Currently listed here.

**Product Team:** Members of the Mautic Product Team. They may also be part of the Core Team, but not necessarily. Currently listed here .

> **Triage Team:** Members of the Mautic Product Team who are responsible for triaging issues and PR’s. They may also be part of the Core Team, but not necessarily. Currently listed here.

Author: This is also broken too

**Closing Stale Pull Requests:** If the PR is pending feedback or inactive for over 30 days, the Product Team may decide to close the PR.

**Closing Stale Issues:** If the issue is pending feedback or inactive for over 14 days, the Product Team may decide to close the issue.

Branching strategy
*******************

> As we maintain multiple major and minor versions of Mautic (3.x, 4.x), we use a defined branching strategy on GitHub.

Reviewer: ```suggestion
As we maintain multiple major and minor versions of Mautic (5.x, 6.x), we use a defined branching strategy on GitHub.
```

Let's bring this more up to date.

> Please refer to the resources in the contributing to Mautic section to understand the branching strategy.

Reviewer: Let's link this to the contributing to Mautic section?

Author: ok Noted

We’ve started to create a supported versions table on GitHub and will update that throughout the year with more specific dates.

> .. note::

Reviewer: This has no link to edit, please remove - I guess it's a copy/paste. Read the Docs should do this automagically for us.

Author: Please, I am not clear. Do you mean I should remove the note ?

> This has no link to edit, please remove - I guess it's a copy/paste. Read the Docs should do this automagically for us.

> We’ve started to create a supported versions table on GitHub and will update that throughout the year with more specific dates.

Reviewer: I replaced this with a link to the mautic.org page at https://mautic.org/releases/ where we maintain the list of supported versions etc.
